### DISCUSSION TITLE
EC2 deployment

### DISCUSSION THREAD

**Discussion by free2fork** [coder/code-server#2677] [Q&A] [answered]
Hi there everyone.
Deployed from Docker into EC2
mkdir -p ~/.config
docker run -it --name code-server -p 127.0.0.1:8080:8080 \
  -v "$HOME/.config:/home/coder/.config" \
  -v "$PWD:/home/coder/project" \
  -u "$(id -u):$(id -g)" \
  -e "DOCKER_USER=$USER" \
  codercom/code-server:latest

Instance is up, container too

~]$ docker ps
CONTAINER ID        IMAGE                         COMMAND                  CREATED             STATUS              PORTS                      NAMES
73ed87b746a1        codercom/code-server:latest   "/usr/bin/entrypoint…"   30 minutes ago      Up 30 minutes       127.0.0.1:8080->8080/tcp   code-server

Curl respond from SSH session 

$ curl -i http://127.0.0.1:8080
HTTP/1.1 302 Found
Location: ./login
Vary: Accept
Content-Type: text/plain; charset=utf-8
Content-Length: 29
Date: Fri, 05 Feb 2021 15:50:56 GMT
Connection: keep-alive

But cant access from https://ec2-example-ip.compute-1.amazonaws.com/

**Comment by oxy** [comment]
Did you configure EC2 to reverse proxy it to Port 80? I haven't used EC2 before, but it looks like that is the most likely reason.

**Reply by free2fork** [reply]
thanks

**Comment by nooobcoder** [comment, marked as answer]
@free2fork please use **-p 8080:8080** to bind it to the default port and not like **-p 127.0.0.1:8080:8080**. Doing so would expose the container on the localhost only and would not be accessible globally.


AN additional information, you can bind an extra volume target to persist the user data on container updates
You need to bind the following volume to a location 
***/home/coder/.local/share/code-server***
As of now, this is not so discussed about, but without this binding, the code-server loses it's data on container update. Binding only the **.config** is not enough.

**BELOW IS THE DOCKER RUN THAT I USE TO DEPLOY MY CODE-SERVER WITH A SELF-SIGNED CERTIFICATE**

_COMMAND TO GENERATE A SELF SIGNED CERTIFICATE_
```
GENERATE CRT AND KEY FILE IN THE CERTS DIRECTORY
openssl req \
       -newkey rsa:2048 -nodes -keyout certs.key \
       -x509 -days 365 -out certs.crt
```

_COMMAND TO DEPLOY THE CONTAINER WITH THE PATH TO THE CERTIFICATES FOR HTTPS_

`sudo docker run -it --name code-server -p 8080:8080 -p 3000:3000  -v "/home/ankurpaul/AppConfig/CodeServer/.config:/home/coder/.config" -v "/home/ankurpaul/AppConfig/CodeServer/data:/home/coder/.local/share/code-server"   -v "/home/ankurpaul/VSCodes:/home/coder/project" -v /home/ankurpaul/AppConfig/CodeServer/certs/certs.crt:/home/coder/cert.crt -v /home/ankurpaul/AppConfig/CodeServer/certs/certs.key:/home/coder/cert.key  -u "$(id -u):$(id -g)"   -e "DOCKER_USER=$USER"  -e "USER=$USER"    codercom/code-server:latest --cert /home/coder/cert.crt --cert-key /home/coder/cert.key`


@free2fork Please mark this discussion as **_answered_** if your query has been solved.
Thank You

**Reply by free2fork** [reply]
YES! Solved. Thank you very much.

Bad idea to create a CloudFormation template to auto deploy it to AWS ?

**Reply by nooobcoder** [reply]
Cloudformation is a very good idea to auto deploy instances and scale them as per your needs. But, be adivsed with auto deployment to prevent racking up of charges.

FYI, I use AWS to host my own code-server instance and it works like a charm, go ahead you should not have any difficulties.
Also, remember to allow port forwarding in your security groups to make development web servers reachable from outside.

Thank You,
Have a great day @free2fork

**Reply by free2fork** [reply]
Perfect! 
Answered.

Cheers!

**Reply by nooobcoder** [reply]
@free2fork  please mark the discussion as answered
Otherwise it would be marked as unanswered.
Thank you
